x/pairing/keeper/unresponsive_provider.go

```diff
@@ -0,0 +1,314 @@
+			return nil
+		}
+		return err
+	}
+
+	// Punish unresponsive providers (the punishment is currently unstake)
+	err = k.punishUnresponsiveProviders(ctx, providersToPunishMap)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Function that returns a map that links between a provider that should be punished and its punishEntry
+func (k Keeper) getUnresponsiveProvidersToPunish(ctx sdk.Context, currentEpoch uint64, epochsNumToCheckCUForUnresponsiveProvider uint64, epochsNumToCheckCUForComplainers uint64) (map[string][]string, error) {
+	// check the epochsNum consts
+	if epochsNumToCheckCUForComplainers <= 0 || epochsNumToCheckCUForUnresponsiveProvider <= 0 {
+		return nil, utils.LavaError(ctx, k.Logger(ctx), "get_unresponsive_providers_to_punish", nil, "epochsNumToCheckCUForUnresponsiveProvider or epochsNumToCheckCUForComplainers are smaller or equal than zero")
+	}
+
+	providersPunishEntryMap := make(map[string]punishEntry) // map of punishEntryKey keys (string) with punishEntry values (punishEntryKey = providerAddress_chainID)
+	providersToPunishMap := make(map[string][]string)       // map of punishEntryKey keys (string) with list of providerPaymentStorage indices values (string array) (punishEntryKey = providerAddress_chainID)
+
+	// Get recommendedEpochNumToCollectPayment
+	recommendedEpochNumToCollectPayment := k.RecommendedEpochNumToCollectPayment(ctx)
+
+	// handle very early epochs (basically, don't punish)
+	epochBlocks, err := k.epochStorageKeeper.EpochBlocks(ctx, currentEpoch)
+	if err != nil {
+		return nil, utils.LavaError(ctx, k.Logger(ctx), "get_epoch_blocks", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", currentEpoch)}, "couldn't get epoch blocks param")
+	}
+	if currentEpoch < (epochsNumToCheckCUForComplainers+epochsNumToCheckCUForUnresponsiveProvider+recommendedEpochNumToCollectPayment)*epochBlocks {
+		// To check for punishment, we have to go back (epochsNumToCheckCUForComplainers+epochsNumToCheckCUForUnresponsiveProvider) epochs from the current epoch. If there isn't enough memory, do nothing
+		return nil, types.EpochTooEarlyForUnresponsiveProviderJailError
+	}
+
+	// Go back recommendedEpochNumToCollectPayment
+	for counter := uint64(0); counter < recommendedEpochNumToCollectPayment; counter++ {
+		previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, currentEpoch)
+		if err != nil {
+			return nil, utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", currentEpoch)}, "couldn't get previous epoch")
+		}
+		currentEpoch = previousEpoch
+	}
+
+	// Go over previous epochs to count the CU serviced by the provider
+	epochAfterCountingProviderCu, err := k.countCuForUnresponsiveProviderAndComplainers(ctx, true, currentEpoch, epochsNumToCheckCUForUnresponsiveProvider, providersPunishEntryMap)
```

Comment:
In Go, maps are always passed as a reference so I don't think this needs changing (see here: https://stackoverflow.com/questions/40680981/are-maps-passed-by-value-or-by-reference-in-go).

I've looked it up because it seemed weird that my code works even when I'm passing the map like this. So by luck, it's ok